src/lib.rs

```diff
@@ -40,188 +46,125 @@
-            KeyValue, SpanKind, StatusCode, TraceContextPropagator, Tracer, Value,
-        },
-        global,
-    },
-    tide::{Middleware, Next, Request, Result},
-    url::Url,
-};
+use std::{convert::TryFrom, net::IpAddr, net::SocketAddr, str::FromStr};
 
-static HTTP_METHOD_ATTRIBUTE: &str = "http.method";
-static HTTP_URL_ATTRIBUTE: &str = "http.url";
-static HTTP_TARGET_ATTRIBUTE: &str = "http.target";
-static HTTP_SCHEME_ATTRIBUTE: &str = "http.scheme";
-static HTTP_STATUS_CODE_ATTRIBUTE: &str = "http.status_code";
-static HTTP_STATUS_TEXT_ATTRIBUTE: &str = "http.status_text";
-static HTTP_FLAVOR_ATTRIBUTE: &str = "http.flavor";
-// TODO: parse UA header
-// static HTTP_USER_AGENT_ATTRIBUTE: &str = "http.user_agent";
-
-static HTTP_HOST_ATTRIBUTE: &str = "http.host";
-static HTTP_SERVER_NAME_ATTRIBUTE: &str = "http.server_name";
-// needs access to the framework's router:
-// static HTTP_ROUTE_ATTRIBUTE: &str = "http.route";
-static HTTP_CLIENT_IP_ATTRIBUTE: &str = "http.client_ip";
-static NET_PEER_IP_ATTRIBUTE: &str = "net.peer.ip";
-static NET_HOST_PORT_ATTRIBUTE: &str = "net.host.port";
-static UNKNOWN: &str = "unknown";
-static EMPTY: &str = "";
+use opentelemetry::{
+    trace::{FutureExt, SpanKind, StatusCode, TraceContextExt, Tracer},
+    Context,
+};
+use opentelemetry_semantic_conventions::trace;
+use tide::{http::Version, Middleware, Next, Request, Result};
+use url::Url;
 
 /// The middleware struct to be used in tide
 #[derive(Default, Debug)]
-pub struct OpenTelemetryTracingMiddleware {
-    _priv: (),
+pub struct OpenTelemetryTracingMiddleware<T: Tracer> {
+    tracer: T,
 }
 
-impl OpenTelemetryTracingMiddleware {
+impl<T: Tracer> OpenTelemetryTracingMiddleware<T> {
     /// Instantiate the middleware
     ///
     /// # Examples
     ///
     /// ```rust,no_run
-    /// // make sure to initialize the tracer first
-    ///
     /// let mut app = tide::new();
-    /// app.with(opentelemetry_tide::OpenTelemetryTracingMiddleware::new());
+    /// let (tracer, _uninstall) = opentelemetry_jaeger::new_pipeline().install().unwrap();
+    /// app.with(opentelemetry_tide::OpenTelemetryTracingMiddleware::new(tracer));
     /// app.at("/").get(|_| async { Ok("Traced!") });
     /// ```
-    pub fn new() -> Self {
-        Self { _priv: () }
+    pub fn new(tracer: T) -> Self {
+        Self { tracer }
     }
 }
 
 #[tide::utils::async_trait]
-impl<State: Clone + Send + Sync + 'static> Middleware<State> for OpenTelemetryTracingMiddleware {
+impl<T: Tracer + Send + Sync, State: Clone + Send + Sync + 'static> Middleware<State>
+    for OpenTelemetryTracingMiddleware<T>
+{
     async fn handle(&self, req: Request<State>, next: Next<'_, State>) -> Result {
-        let mut req_headers = std::collections::HashMap::new();
-        for (k, v) in req.iter() {
-            req_headers.insert(k.to_string(), v.last().to_string());
+        let method = req.method().clone();
```

Comment:
I think it wants us to do something like
```suggestion
        let method = *req.method();
```

This should be okay, because we copy the data into new strings anyway.
Another way would be to use `req.method()` explicitly in the 2 places below and get rid of this temporary variable.

Reply:
Didn't realise Method was `Copy`